### PR TITLE
Implementing HappyBase Table atomic counter helpers.

### PR DESCRIPTION
**Pull request by dhermes**

### CODE REVIEW

File: gcloud/bigtable/happybase/table.py [hunk 1]
```diff
@@ -127,6 +130,88 @@ def regions(self):
...
+        :type value: int
+        :param value: Amount to increment the counter by. (If negative,
+                      this is equivalent to decrement.)
+
+        :rtype: int
+        :returns: Counter value after incrementing.
+        """
+        row = self._low_level_table.row(row)
+        if isinstance(column, six.binary_type):
+            column = column.decode('utf-8')
```
Comment (dhermes) [inline gcloud/bigtable/happybase/table.py]: @tseaver I added this when moving the code over, but maybe it's too defensive? I wanted to ensure `column` was the right type before `column.split(':')`. WDYT?

**Comment by dhermes** [comment]
@jgeewax HappyBase also [has a `counter_set`][2] but as of now, I [had made it `NotImplemented`][1] because Bigtable doesn't have a way to make it atomic. WDYT is the right move here? I did make a note in the docstring of `counter_get` that

```
.. note::

    Application code should **never** store a counter value directly;
    use the atomic :meth:`counter_inc` and :meth:`counter_dec` methods
    for that.
```

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/8d91feb7daf4ccac7818d468249c4c95f0e14a5e/gcloud_bigtable/happybase/table.py#L793-L815
[2]: https://github.com/wbolster/happybase/blob/9cbd718c10a3089f234f1eac1236b631e1f8e7cd/happybase/table.py#L525

**Comment by tseaver** [comment]
@dhermes, @jgeewax:  I'm pretty much offline until Thursday, 2016-02-25 (in Palo Alto teaching a course).  Likewise for #1515, #1516, #1502, #1512, and #1513.

File: gcloud/bigtable/happybase/table.py (hunk 1 above)
Comment (jonparrott) [inline gcloud/bigtable/happybase/table.py]: If there are cases where column isn't a unicode, then I think it's fair to have that here.

**Comment by jonparrott** [comment]
This LGTM, although from a readability nit standpoint `counter_inc` is a little hard to follow at first glance. I'm unsure why.

File: gcloud/bigtable/happybase/table.py (hunk 1 above)
Comment (dhermes) [inline gcloud/bigtable/happybase/table.py]: The `column` variable is passed by the user so I was allowing for them to do as they please. The HappyBase concept of `fam:` or `fam:col` is different from that of Cloud Bigtable.

In Cloud Bigtable the fields are separate: [`Column.qualifier`][1] (equivalent of `col` in `fam:col`) is `bytes` and [`Family.name`][2] (i.e. `fam`) is a string.

It's unclear what the right thing here is. For example if `fam` is a string but `col` has non-ascii bytes, then decoding to `utf-8` may be a bad idea.

[1]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/7c77c170c8818a0440e0d2f4ecedcb8329c1f8d9/bigtable-protos/src/main/proto/google/bigtable/v1/bigtable_data.proto#L58
[2]: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/7c77c170c8818a0440e0d2f4ecedcb8329c1f8d9/bigtable-protos/src/main/proto/google/bigtable/v1/bigtable_data.proto#L45

Comment (jonparrott) [inline gcloud/bigtable/happybase/table.py]: Yikes.

**Comment by dhermes** [comment]
RE: "Hard to follow." That's the kind of feedback I was hoping to surface (by breaking into small PRs). Do you think the low-level row should handle more of the responsibilities? Do you think the method should use a helper? etc.

**Comment by jonparrott** [comment]
Perhaps just some comments in-line. Poor example, but what is `bytes_value = column_cells[0][0]` doing exactly? Why `[0][0]`?

**Comment by jonparrott** [comment]
(this also just may be my unfamiliarity with bigtable showing)

File: gcloud/bigtable/happybase/table.py [hunk 2]
```diff
@@ -127,6 +130,88 @@ def regions(self):
...
+        column_family_id, column_qualifier = column.split(':')
+        row.increment_cell_value(column_family_id, column_qualifier, value)
+        modified_cells = row.commit_modifications()
+        column_cells = modified_cells[column_family_id][column_qualifier]
+        if len(column_cells) != 1:
+            raise ValueError('Expected server to return one modified cell.')
+        bytes_value = column_cells[0][0]
```
Comment (dhermes) [inline gcloud/bigtable/happybase/table.py]: Per @jonparrott, this should have a comment to explain what `[0][0]` is doing.

**Comment by dhermes** [comment]
I don't think it's your unfamiliarity, but that also is good in that you don't have a bias towards how things should be. I had to look up what `[0][0]` is for: `Row.commit_modifications` returns a dictionary keyed by column families. Each column family contains a dictionary keys by column names (within the family). Each column has a list of tuples. Each tuple is a pair of the bytes in the cell and the associated timestamp. It's enough to need a breather after the explanation.

**Comment by jonparrott** [comment]
Wow, that is a lot to explain.

**Comment by dhermes** [comment]
Luckily the [docstring][1] proves that a picture is worth a thousand words:

```python
{
    u'col-fam-id': {
        b'col-name1': [
            (b'cell-val', datetime.datetime(...)),
            (b'cell-val-newer', datetime.datetime(...)),
        ],
        b'col-name2': [
            (b'altcol-cell-val', datetime.datetime(...)),
        ],
    },
    u'col-fam-id2': {
        b'col-name3-but-other-fam': [
            (b'foo', datetime.datetime(...)),
        ],
    },
}
```

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/115263bbabd40ae6cb1480e6b4cf4dea94f3f052/gcloud/bigtable/row.py#L477-L494

**Comment by jonparrott** [comment]
Ah, very helpful. Can you include a comment that notes the expected format of the `column_cells` object?

**Comment by dhermes** [comment]
@jonparrott PTAL

**Comment by jonparrott** [comment]
LGTM, please squash commits.
